### PR TITLE
ActionRequest's javadoc should be more explicit about the need to implement IndicesRequest

### PR DESCRIPTION
**Pull request by dweiss**
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

**Comment by dweiss** [comment]
This PR is just a suggestion, but I found it quite hard (down to source-code analysis) to figure out why my custom action is being rejected by x-pack. Turned out I wasn't implementing this interface (which isn't mentioned anywhere I think). A link or cross-reference here would be very handy now that x-pack is open.

File: server/src/main/java/org/elasticsearch/action/ActionRequest.java [hunk 1]
```diff
@@ -25,6 +25,14 @@
 
 import java.io.IOException;
 
+/**
+ * Note: classes that extend the {@link ActionRequest} should also implement
+ * {@link IndicesRequest} to provide index access information for x-pack security
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/action/ActionRequest.java]: We strongly prefer to not mention X-Pack in the Apache Licensed codebase.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I have no opinion on the overall change but I did leave one comment about keeping the OSS source agnostic to the existence of the X-Pack source.

**Comment by jasontedor** [comment]
Upon further reflection, this change would not be correct. Our codebase itself contains many action requests that do not implement `IndicesRequest`; look at #28671 for one of many examples. Thanks for the PR but I am going to close this without merging.

**Comment by dweiss** [comment]
Well, without it the action will not pass through x-pack security filters, so either the logic there is wrong or something should be done to help people work with x-pack, even if it's separate from ASF code. I spent two days trying to get my plugin work with x-pack -- it's a nightmare when you have to resort to stepping through the code in the authentication filters to see why your action is being rejected.

**Comment by dweiss** [comment]
btw. RestoreSnapshotAction is an admin request, the action I talk about is an index-related request. If it doesn't implement IndicesRequest, it just won't work and fail with a somewhat cryptic message.

**Comment by s1monw** [comment]
maybe we can rephrase it to something like this:

```Java
/**
 * Note: classes that extend the {@link ActionRequest} that operate on one or more indices 
 * should also implement {@link IndicesRequest} to provide index access information for further processing
 *
 * @see IndicesRequest
 * @see IndicesRequest.Replaceable
 */
```

I don't think it's x-pack specific per-se

**Comment by dweiss** [comment]
Alternatively, something like "a short guide for x-pack compatible action developers" inside the x-pack repository would be really handy. Now that the two codebases coexist I don't think it's that bad to have cross-references, actually, but if you don't want them then helping out with conventions you use on the open source side (and exploit in x-pack) would be very handy indeed. The two I can think about immediately are action name prefixes (for permission grants) and Action-IndicesRequest relationship, which isn't at all obvious. 

I can't stress this enough -- it's not me complaining, it's people taking our plugin (which worked fine with ES), installing x-pack and then asking us why it no longer works... For which we, until not so long ago, could only shrug and say "we don't know".

**Comment by s1monw** [comment]
@jaymode can you take this on you ^^

**Comment by jaymode** [comment]
@s1monw sure, I will pick this up. @albertzaharovits recently published a blog about x-pack compatibility specifically geared at plugin authors https://www.elastic.co/blog/plugins-alongside-the-x-pack-security-plugin-for-elasticsearch. We can incorporate that information back into the repository and expand on it as needed.

**Comment by dweiss** [comment]
For what it's worth, this blog post entry doesn't mention the IndicesRequest interface. Like I said -- this interface is essential for actions with the "indices" prefix, which goes back to my original point I tried to make on this PR.

**Pull request closed without merging**